### PR TITLE
Restructure project in multiple crates

### PR DESCRIPTION
**Pull request by Istar-Eldritch**
This PR relates to #35 .

@z2oh So far I just created a multiple crate structure. After checking the code base I believe it's already organized enough as to do the complete restructure in this single PR. I'll set up the restructuring of the different parts in independent commits to simplify its review.

Let me know if this is something you disagree with. I plan to start working on such refactor tomorrow evening UTC time.

### CODE REVIEW

**Comment by z2oh** [comment]
I was thinking more along the lines of having the root crate be the interface, and then the expression and parser libraries being sub crates. I believe this is along the same lines as [ripgrep](https://github.com/BurntSushi/ripgrep)'s organization, but admittedly I do not know which way will work best :)

Ideally, I want to be able to type `cargo run` at the _root_ directory and have it run. If the organizational strategy proposed in this PR accomplishes that, then it is fine with me :)

**Comment by Istar-Eldritch** [comment]
It was doing that, but I agree. Let's do it as you advise. :wink:

**Comment by Istar-Eldritch** [comment]
@z2oh the main changes are done. 

There is a flag I would like to raise regarding the [`nom` public export](https://github.com/z2oh/sexe/pull/36/files#diff-fc7ff63890d2ac626ab17a006fa4fb40R5) from the parser crate. This, while handy has some serious drawbacks regarding maintainability and suppor of the crate interface. For example, if the nom developers decide to change this type in the future the `sexe_parser` crate will be forced to update it's interface too, making it hightly coupled to `nom`. This is probably a discussion for another time so I just wanted to mention ensure awareness.

**Comment by Istar-Eldritch** [comment]
Also, I just realised it now by looking at the diff. I use `rustfmt` automatically and it forced the standard style in all the files I touched :confounded: . Let me know if this is an issue and I'll recover their previous state.

File: src/interface.rs [hunk 1]
```diff
@@ -113,33 +110,41 @@ enum Error {
     RangeError,
 }
 
-fn evaluate_function_over_domain(start_x: f64, end_x: f64, resolution: u32, function_string: &str) -> Result<Vec<(f64, f64)>, Error> {
+fn evaluate_function_over_domain(
```
Comment (z2oh) [inline src/interface.rs:113]: Could you possibly refactor this function as part of the rearchitecture?

This function makes the most sense to be a publicly exported function in the `sexe_expression` crate. We should also change this from taking `function_string` to an actual expression node. This means that where we call this function, we will need to first parse the function string, then pass the expression node to _this_ function in the expression module.

**Review by z2oh — CHANGES_REQUESTED** [review]
Don't worry about rustfmt, it needs to be done sooner or later 🙂

Would you mind moving the `evaluate_function_over_domain` function to the expression crate? Thanks for rearchitecting this, awesome work!! 😀

**Comment by z2oh** [comment]
We can fix the coupling with nom by adding a function to the `parser` crate which takes a `&str` and returns an `ExpressionNode`. This way we won't need to deal with `CompleteStr` except for inside the `parser` crate. This changes goes hand in hand with the change I requested above about moving the `evaluate_function_over_domain` function to the `expression` crate.

If we do that, we should be able to remove the re-export of `CompleteStr` right?

**Comment by z2oh** [comment]
Closed via #37

**Comment by Istar-Eldritch** [comment]
Damn, sorry @z2oh I completely forgot about this.

**Comment by z2oh** [comment]
No problem @Istar-Eldritch ! Thanks for doing the legwork on this! 😀

**Pull request merged**